### PR TITLE
Improve startup error handling

### PR DESCRIPTION
**Pull request by rorp**

### CODE REVIEW

File: eclair-core/src/main/resources/reference.conf [hunk 1]
```diff
@@ -23,6 +23,7 @@ eclair {
     rpcuser = "foo"
     rpcpassword = "bar"
     zmq = "tcp://127.0.0.1:29000"
+    max-unsync-blocks = 2147483647 // for low memory nodes like Raspbery Pi set this parameter to a small number, eg 2, 1, or even 0
```
Comment (sstone) [inline eclair-core/src/main/resources/reference.conf]: Why such a high value ? if your bitcoind has more headers than blocks then it's extremely likely that it's out of sync. I'd rather have a much smaller default (1 or 2)

File: eclair-core/src/main/scala/fr/acinq/eclair/Setup.scala [hunk 2]
```diff
@@ -98,22 +99,26 @@ class Setup(datadir: File,
...
+      val (progress, chainHash, bitcoinVersion, unspentAddresses, blocks, headers) = await(future, 30 seconds, "bicoind did not respond after 30 seconds")
+      assert(Version(bitcoinVersion) >= Version("0.16.0"), "Eclair requires Bitcoin Core 0.16.0 or higher")
       assert(chainHash == nodeParams.chainHash, s"chainHash mismatch (conf=${nodeParams.chainHash} != bitcoind=$chainHash)")
       if (chainHash != Block.RegtestGenesisBlock.hash) {
         assert(unspentAddresses.forall(address => !isPay2PubkeyHash(address)), "Make sure that all your UTXOS are segwit UTXOS and not p2pkh (check out our README for more details)")
       }
-      assert(progress > 0.99, "bitcoind should be synchronized")
+      val maxUnsyncBlocks = config.getInt("bitcoind.max-unsync-blocks")
+      logger.info(s"progress=$progress blocks=$blocks headers=$headers, maxUnsyncBlocks=$maxUnsyncBlocks")
+      assert(progress > 0.999 && headers - blocks <= maxUnsyncBlocks, "bitcoind should be synchronized")
```
Comment (sstone) [inline eclair-core/src/main/scala/fr/acinq/eclair/Setup.scala]: I'd rather have 2 separate tests: one for `progress`, and the other one the `headers - blocks`. This last check should be deactivated on testnet, and while you're at it include the `progress` , `headers` and ` blocks` values in the log message :)

File: eclair-core/src/main/scala/fr/acinq/eclair/Setup.scala [hunk 3]
```diff
@@ -98,22 +99,26 @@ class Setup(datadir: File,
...
             .filter(value => (value \ "spendable").extract[Boolean])
             .map(value => (value \ "address").extract[String])
         }
-      } yield (progress, chainHash, bitcoinVersion, unspentAddresses)
+      } yield (progress, chainHash, bitcoinVersion, unspentAddresses, blocks, headers)
       // blocking sanity checks
-      val (progress, chainHash, bitcoinVersion, unspentAddresses) = Await.result(future, 30 seconds)
-      assert(bitcoinVersion.startsWith("16"), "Eclair requires Bitcoin Core 0.16.0 or higher")
+      val (progress, chainHash, bitcoinVersion, unspentAddresses, blocks, headers) = await(future, 30 seconds, "bicoind did not respond after 30 seconds")
+      assert(Version(bitcoinVersion) >= Version("0.16.0"), "Eclair requires Bitcoin Core 0.16.0 or higher")
```
Comment (sstone) [inline eclair-core/src/main/scala/fr/acinq/eclair/Setup.scala]: There is already a PR (#682) to support bitcoind 0.17 which behaves a bit differently from 0.16, so I think it's better to keep the old version check in this PR.

File: eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/zmq/ZMQActor.scala [hunk 4]
```diff
@@ -47,20 +48,26 @@ class ZMQActor(address: String, connected: Option[Promise[Boolean]] = None) exte
   implicit val ec: ExecutionContext = context.system.dispatcher
 
   // we check messages in a non-blocking manner with an interval, making sure to retrieve all messages before waiting again
-  def checkEvent: Unit = Option(Event.recv(monitor, ZMQ.DONTWAIT)) match {
-    case Some(event) =>
-      self ! event
-      checkEvent
+  def checkEvent: Unit = checkEventRec
+
+  @tailrec
```
Comment (sstone) [inline eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/zmq/ZMQActor.scala]: Thanks! it should fix #698

File: eclair-core/src/main/scala/fr/acinq/eclair/utils/Version.scala [hunk 5]
```diff
@@ -0,0 +1,21 @@
+package fr.acinq.eclair.utils
+
+import scala.util.Try
+
+case class Version(value: String) extends Ordered[Version] {
```
Comment (sstone) [inline eclair-core/src/main/scala/fr/acinq/eclair/utils/Version.scala]: See above, I think we should keep the old version check in this PR

**Comment by Christewart** [comment]
Perhaps break out ZMQ ports like they do in `lnd`? I.e. different ports for `rawblock`, `rawtx` etc?

File: eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/zmq/ZMQActor.scala [hunk 6]
```diff
@@ -47,26 +48,33 @@ class ZMQActor(address: String, connected: Option[Promise[Boolean]] = None) exte
...
 
-  checkEvent
-  checkMsg
+  self ! 'checkEvent
+  self ! 'checkMsg
 
   override def receive: Receive = {
+    case 'checkEvent =>
+      checkEvent
+      context.system.scheduler.scheduleOnce(1 second, self ,'checkEvent)
```
Comment (rorp) [inline eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/zmq/ZMQActor.scala]: @sstone I'm just curious, what are the benefits of scheduling a queue check this way?

**Comment by rorp** [comment]
> Perhaps break out ZMQ ports like they do in `lnd`? I.e. different ports for `rawblock`, `rawtx` etc?

Maybe. But it's out of scope of this PR. What I'm trying to do here is to let the users see not just StackOverflowException's or OutOfMemoryError's, but what actually cause them.

**Pull request merged**